applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -372,253 +427,158 @@ void File::AddPath(const std::string& rPath)
+
     KRATOS_INFO_IF("HDF5Application", GetEchoLevel() == 2)
         << "Write time \"" << rObjectPath << '/' << rName
         << "\": " << timer.ElapsedSeconds() << std::endl;
-    KRATOS_CATCH("Path: \"" + rObjectPath + '/' + rName + "\".");
-}
 
-void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, const array_1d<double, 3>& rValue)
-{
-    KRATOS_TRY;
-    Vector<double> vector_value = rValue;
-    WriteAttribute(rObjectPath, rName, vector_value);
     KRATOS_CATCH("Path: \"" + rObjectPath + '/' + rName + "\".");
 }
 
-void File::WriteDataSet(const std::string& rPath, const Vector<int>& rData, WriteInfo& rInfo)
+void File::WriteAttribute(
+    const std::string& rObjectPath,
+    const Parameters Attributes)
 {
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
+    KRATOS_TRY
 
-void File::WriteDataSet(const std::string& rPath, const Vector<double>& rData, WriteInfo& rInfo)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
-
-void File::WriteDataSet(const std::string& rPath, const Vector<array_1d<double, 3>>& rData, WriteInfo& rInfo)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
+    for (auto itr = Attributes.begin(); itr != Attributes.end(); ++itr) {
+        const auto attrib_name = itr.name();
 
-void File::WriteDataSet(const std::string& rPath, const Matrix<int>& rData, WriteInfo& rInfo)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
+        if (itr->IsInt()) {
+            WriteAttribute(rObjectPath, attrib_name, itr->GetInt());
+        } else if (itr->IsDouble()) {
+            WriteAttribute(rObjectPath, attrib_name, itr->GetDouble());
+        } else if (itr->IsString()) {
+            WriteAttribute(rObjectPath, attrib_name, itr->GetString());
+        } else if (itr->IsMatrix()) {
+            WriteAttribute(rObjectPath, attrib_name, itr->GetMatrix());
+        } else if (itr->IsVector()) {
+            WriteAttribute(rObjectPath, attrib_name, itr->GetVector());
+            bool is_int_array = true;
+            bool is_double_array = true;
 
-void File::WriteDataSet(const std::string& rPath, const Matrix<double>& rData, WriteInfo& rInfo)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
-
-void File::WriteDataSetIndependent(const std::string& rPath, const Vector<int>& rData, WriteInfo& rInfo)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
+            for (const auto& v : *itr) {
```

Comment:
```suggestion
            for (auto v : *itr) {
```

Reply:
I would keep it const, since it gives the indication for the reader that it is not being mutated.

Reply:
then

```suggestion
            for (const auto v : *itr) {
```

my main beef is with the ref (`Parameters` is **not** a container)

Reply:
I would like to avoid a copy operation, and I don't see a problem by using the reference in this case since, we read the underlying data and disregard the itr.  We don't use it anymore. I think, in this case this is safe and we avoid copy operations.